core/src/main/scala-2/chisel3/Aggregate.scala

```diff
@@ -0,0 +1,314 @@
+    m:   Int,
+    p:   Int
+  )(gen: (Int, Int, Int) => T
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Vec[Vec[Vec[T]]] = _tabulateImpl(n, m, p)(gen)
+
+  /** Creates a new [[Vec]] of length `n` composed of the result of the given
+    * function applied to an element of data type T.
+    *
+    * @param n number of elements in the vector
+    * @param gen function that takes in an element T and returns an output
+    * element of the same type
+    */
+  def fill[T <: Data](n: Int)(gen: => T): Vec[T] = macro VecTransform.fill
+
+  /** @group SourceInfoTransformMacro */
+  def do_fill[T <: Data](n: Int)(gen: => T)(implicit sourceInfo: SourceInfo): Vec[T] = _fillImpl(n)(gen)
+
+  /** Creates a new 2D [[Vec]] of length `n by m` composed of the result of the given
+    * function applied to an element of data type T.
+    *
+    * @param n number of inner vectors (rows) in the outer vector
+    * @param m number of elements in each inner vector (column)
+    * @param gen function that takes in an element T and returns an output
+    * element of the same type
+    */
+  def fill[T <: Data](n: Int, m: Int)(gen: => T): Vec[Vec[T]] = macro VecTransform.fill2D
+
+  /** @group SourceInfoTransformMacro */
+  def do_fill[T <: Data](
+    n:   Int,
+    m:   Int
+  )(gen: => T
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Vec[Vec[T]] = _fillImpl(n, m)(gen)
+
+  /** Creates a new 3D [[Vec]] of length `n by m by p` composed of the result of the given
+    * function applied to an element of data type T.
+    *
+    * @param n number of 2D vectors inside outer vector
+    * @param m number of 1D vectors in each 2D vector
+    * @param p number of elements in each 1D vector
+    * @param gen function that takes in an element T and returns an output
+    * element of the same type
+    */
+  def fill[T <: Data](n: Int, m: Int, p: Int)(gen: => T): Vec[Vec[Vec[T]]] = macro VecTransform.fill3D
+
+  /** @group SourceInfoTransformMacro */
+  def do_fill[T <: Data](
+    n:   Int,
+    m:   Int,
+    p:   Int
+  )(gen: => T
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Vec[Vec[Vec[T]]] = _fillImpl(n, m, p)(gen)
+
+  /** Creates a new [[Vec]] of length `n` composed of the result of the given
+    * function applied to an element of data type T.
+    *
+    * @param start First element in the Vec
+    * @param len Lenth of elements in the Vec
+    * @param f Function that applies the element T from previous index and returns the output
+    * element to the next index
+    */
+  def iterate[T <: Data](start: T, len: Int)(f: (T) => T): Vec[T] = macro VecTransform.iterate
+
+  /** @group SourceInfoTransformMacro */
+  def do_iterate[T <: Data](
+    start: T,
+    len:   Int
+  )(f:     (T) => T
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Vec[T] = _iterateImpl(start, len)(f)
+}
+
+/** A trait for [[Vec]]s containing common hardware generators for collection
+  * operations.
+  */
+trait VecLike[T <: Data] extends VecLikeImpl[T] with SourceInfoDoc {
```

Comment:
@adkian-sifive I ended up going back and adding VecLikeImpl so no need to document the previous weirdness here.